common/log/interface.go

```diff
@@ -0,0 +1,90 @@
+package log
+
+// Fields is a K-V map for logging tags
+type Fields map[tagType]interface{}
+
+// Logger is our abstraction for logging
+// Usage examples:
+//  1) logger = logger.WithFields(log.Fields{
+//          log.TagWorkflowEventID: 123,
+//          log.TagDomainID       : "test-domain-id"})
+//     logger.Info("hello world")
+//  2) logger.Info("hello world", log.Fields{
+//          log.TagWorkflowEventID : 123,
+//          log.TagDomainID        : "test-domain-id",
+//	   })
+//  Note: msg should be static, it is not recommended to use fmt.Sprintf() for msg.
+//        Anything dynamic should be tagged.
+type Logger interface {
+	Debug(msg string, tags ...Fields)
+	Info(msg string, tags ...Fields)
+	Warn(msg string, tags ...Fields)
+	Error(msg string, tags ...Fields)
+	Fatal(msg string, tags ...Fields)
+	WithFields(tags Fields) Logger
+
+	// We provide shortcuts for logging component lifecycle logs
+	SetComponent(component tagValueTypeSysComponent) Logger
+	DebugWithLifecycle(msg string, lifecycle tagValueTypeSysLifecycle, tags ...Fields)
+	InfoWithLifecycle(msg string, lifecycle tagValueTypeSysLifecycle, tags ...Fields)
+	WarnWithLifecycle(msg string, lifecycle tagValueTypeSysLifecycle, tags ...Fields)
+	ErrorWithLifecycle(msg string, lifecycle tagValueTypeSysLifecycle, tags ...Fields)
+	FatalWithLifecycle(msg string, lifecycle tagValueTypeSysLifecycle, tags ...Fields)
+}
+
+// we intentionally make tagType module-private to enforce using pre-defined tag keys
+type tagType struct {
+	name      string
+	valueType valueTypeEnum
+}
+
+type valueTypeEnum string
```

Comment:
Why use a string to represent the enum instead of the more common pattern of using an int?

Reply:
Because that value will be actually passed into ELK and need to be human readable. One bad example is our workflow-event(the name is confusing, we will call it component lifecycle). It use int and turns out super painful for us to search in ELK. See https://github.com/uber/cadence/blob/master/common/logging/events.go#L31
Another good example is Chronos, after Venkat's refactoring, we are using string as value